frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
+      reUsePorts = am.getStatusManager().getLiveAssociatedContainerPorts(taskRoleName);
+      if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && requestedPortsCache.containsKey(taskRoleName)) {
+        reUsePorts = requestedPortsCache.get(taskRoleName);
+        // the cache only guide the next task to use previous requesting port.
+        requestedPortsCache.remove(taskRoleName);
       }
+    }
+    SelectionResult result = select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, reUsePorts);
+    if (!requestedPortsCache.containsKey(taskRoleName) && pendingTaskNumber > 1)
+      requestedPortsCache.put(taskRoleName, result.getOptimizedResource().getPortRanges());
+    return result;
+  }
 
-      LOGGER.logWarning(
-          "select: The Request will be relaxed to RM");
+  @VisibleForTesting
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType, int pendingTaskNumber) throws NotAvailableException {
+    return select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, null);
+  }
+
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType,
+      int pendingTaskNumber, List<ValueRange> reUsePorts) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: Request: Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s], TaskNumber: [%d], ReUsePorts: [%s]",
+        requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, ValueRangeUtils.toString(reUsePorts));
+
+    randomizeNodes();
+    filterNodesByNodeLabel(requestNodeLabel);
+    filterNodesByGpuType(requestNodeGpuType);
+    if (!am.getConfiguration().getLauncherConfig().getAmAllowNoneGpuJobOnGpuNode()) {
+      int jobTotalRequestGpu = am.getRequestManager().getTotalGpuCount();
+      filterNodesForNonGpuJob(jobTotalRequestGpu);
+    }
+
+    ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
+    if (ValueRangeUtils.getValueNumber(reUsePorts) > 0) {
+      LOGGER.logInfo(
+          "select: re-use pre-allocated ports: [%s]", ValueRangeUtils.toString(reUsePorts));
+      optimizedRequestResource.setPortRanges(reUsePorts);
     }
 
+    filterNodesByResource(optimizedRequestResource, am.getConfiguration().getLauncherConfig().getAmSkipLocalTriedResource());
+
+    filterNodesByRackSelectionPolicy(optimizedRequestResource, pendingTaskNumber);
+    if (filteredNodes.size() < 1) {
+      //don't have candidate nodes for this request.
+      if (requestNodeGpuType != null || requestResource.getPortNumber() > 0) {
+        //If gpuType or portNumber is specified, abort this request and try later.
+        throw new NotAvailableException(String.format("Don't have enough nodes to fit in optimizedRequestResource:%s, NodeGpuType: [%s]",
```

Comment:
>NotAvailableException [](start = 18, length = 21)

This exeception may not caused by optimizedRequestResource, such as no nodelabel matched.
You should just refer previous implement:
    if (selectionResult != null) {
      LOGGER.logInfo(
          "select: Found a SelectionResult satisfies the Request: SelectionResult: [%s]",
          selectionResult);
    } else {
      LOGGER.logWarning(
          "select: Cannot found a SelectionResult satisfies the Request: " +
              "Check whether the Request can be relaxed to RM");
      String notRelaxLogPrefix = "select: The Request cannot be relaxed to RM: Reason: ";

      // Test NodeGpuType
      if (requestNodeGpuType != null) {
        throw new NotAvailableException(
            String.format(notRelaxLogPrefix +
                    "NodeGpuType is specified: Request NodeGpuType: [%s]",
                requestNodeGpuType));
      }

      LOGGER.logWarning(
          "select: The Request will be relaxed to RM");
    }